### ISSUE TITLE
Significant Speedup of Kanji Search

### ISSUE THREAD

**Issue by Garethp** [birtles/rikaichamp#43]
So, I noticed Kanji Searches in RikaiRebuilt was actually fairly slow. Makes sense, every time it needs to find information for the Kanji it searches the 12,500 line string line by line looking for the Kanji. After taking a look at what it did, it was basically just returning whatever line starts with the kanji given. So I managed to replace that function with something much more readable and faster. Maybe you want to test it and see how you like it.

My Solution:
```js
if (!this.kanjiData) {
    this.kanjiData = await FileReader.read('../resources/kanji.dat')
        .then(response => response.text())
        .then(data => data.split('\n').filter(line => line.length !== 0))
        .then(array => array.reduce((obj, current) => {
            obj[current.charAt(0)] = current;
            return obj;
        }, {}));
}

kde = this.kanjiData[character];
if (!kde) return null;

```

So, pretty simply, rather than just read the entire file in to string and then call `this.find` on that string I load the string up, split it in to an array and then reduce the array in to an obj where the key is the kanji character. Bam, instant indexing for searching. Seems much much faster for me. I'll probably take it a bit further at some point and pre-prepare the actual result (because the `kanjiSearch` function is ugly as hell) and maybe even write the entire thing as a JSON so there's no need to process on the client at all. But for now, this should be useful for you as well

**Comment by birtles** [comment]
Hi Gareth!

Thanks for the issue! I am (slowly) working on a completely different database format that uses IndexedDB native indices for the word lookup so I'm less inclined at the moment to make improvements to the old format. I'd be interested in exploring how we can work together, though. Is there an IRC / Slack / email / Gitter or something I can contact you on? My email is birtles@gmail.com if email works for you.

**Comment by Garethp** [comment]
Well, this is specifically for the `kanjiSearch` as opposed to the `wordSearch`, but if you were planning on rolling the Kanji Search in to your IndexedDb format, then fair enough

**Comment by birtles** [comment]
This should be fixed in 0.1.2 which moves the kanji data to IDB.
